### DISCUSSION TITLE
test/usb-serial-cdc-gadget and Windows / USB mass-storage gadget development

### DISCUSSION THREAD

**Discussion by messinm** [rsta2/circle#455] [General]
Was trying test\usb-serial-cdc-gadget to connect Raspi zero2 to Windows (11), and could not get it to connect properly. Code 10, device would not start. 

I have a board running Yocto with gadget driver,and it connected to Windows with no problems, and so I ran packet captures to compare the two. The Yocto driver responds to GET DESCRIPTOR DEVICE with bDeviceClass =0x02 (CDC), while Circle responds with 0x0.

I changed CUSBCDCGadget::s_DeviceDescriptor so that Circle would respond with bDeviceClass =0x02, and that fixed the issue. Seems like this would work with other OS as well, but I haven't tested.

Should I create an issue for this?

**Comment by rsta2** [comment]
The serial USB CDC gadget has be developed by @sebastienNEC. Sebastien do you 
think, the described modification should be generally added for the device 
descriptor?

I think, we can discuss this here.

**Comment by sebastienNEC** [comment]
Aha indeed, I tested your modification and linux (5.4.0) still works, while Windows (10) now works (it didn't use to).

This needs fixed in the develop branch indeed, sorry for overlooking that as I had only tested on linux !

**Reply by messinm** [reply]
No worries, it was fun to dig into low-level USB stuff (I don't have any expertise in USB). Having read a bit more, seems that bDeviceClass=0x0 indicates a composite device. Did you select this so that additional functions could be added to your driver?

I've been thinking it would be nice to add USB mass storage to the gadget USB, to allow easy access to the RPi Sd card. Might start experimenting with that, any suggestions would be welcome.

**Comment by sebastienNEC** [comment]
While I'm playing with USB: would it be reasonable to replace in dwusbgadget.cpp all CTimer::Get()->xxxxDelay() by their CScheduler::Get()->xxxSleep() equivalent ?

I tried that for the CDC gadget and it works, of course by having the USB initialization code in a CTask.

If so, care must be taken to fallback to CTimer delays if the scheduler was not instantiated. Noting that one cannot use a test on (CScheduler::Get() == nullptr) because the assert would fail.

**Comment by sebastienNEC** [comment]
To be honest, the bDeviceClass=0x0 was an oversight on my side, but indeed, a composite CDC/MSC device would be super useful.

From my experience on the STM32 though, windows/linux are pretty fussy about the exact descriptors of composite devices - they only recognize certain configuration. Do you have an existing composite device whose descriptor you could copy ? 

Now mind you, this was a few years ago, the situation might have improved :)

**Reply by messinm** [reply]
No, I don't have such a device. https://docs.kernel.org/usb/gadget_multi.html has some interesting information. I might be able to use another board with Linux and it's composite gadget for testing.

**Reply by sebastienNEC** [reply]
Sounds good ! I would however start with a simple MSC service to first figure out the interaction with the rest of circle. In particular, how to link to the actual filesystem or SD card, and how to harmonize with the existing usbmassdevice.cpp
I remember that for the CDC service, Rene had to make some changes to the Host side of things to harmonize with the Gadget side...

**Reply by rsta2** [reply]
I think, there is no need to harmonize the MSC gadget driver with *usbmassdevice.cpp*. This is different from the serial CDC gadget device, which shares an API with the USB serial host device, because the serial interface is bidirectional. The mass-storage device basically allows to read and write 512-byte data blocks, which are addressed with a block number. It should be enough to pass a `CDevice` pointer to a block device instance to the constructor of the gadget driver, so that it knows, which device to access.

**Reply by messinm** [reply]
Ok, I'll plan to work with a CDevice. I've scanned through circle to try to see how the SD card would be accessed as a CDevice, don't see an example, can you point me to one? I know it must be there and I've missed it. I saw addon/SDCard where it appears that it is not accessed as a CDevice.

Also, have a question about endpoint 0 in the mass storage gadget - It appears (from packet capture of a mem stick) that there is bRequest=0xfe "GET MAX LUN" that is sent by host to device on endpoint 0, and the dwusbgadgetendpoint0 does not appear to handle this (makes sense since this request is specific to mass storage). So, will I need to create an endpoint0 class for the mass storage gadget? Perhaps subclass dwusbgadgetendpoint0 and delegate back to it except for the 0xfe request?

Thanks.

**Reply by rsta2** [reply]
`CEMMCDevice` is derived from the base class `CDevice`, so you can use a pointer to `CEMMCDevice` everywhere where a `CDevice` pointer is expected. Have a look at *sample/07-usbstorage*, which directly accesses the Master Boot Record of an USB flash drive. The low-level access to `CEMMCDevice` is similar.

Currently USB vendor requests to EP0 are simply ignored. There must be a method, which allows to implement vendor requests in the device specific gadget driver, so that `GET MAX LUN` can be implemented. I will deal with this next week.

There is another thing, which is not implemented yet, and probably needs to be provided. The `GetSize()` method is not implemented for `CEMMCDevice` so far. It requires to parse the CID structure of the SD card, which is non-trivial. The EMMC driver, which I ported to Circle, did not implement it, and it was not needed in Circle so far. The class `CUSBBulkOnlyMassStorageDevice` implements is for USB flash drives.

**Reply by rsta2** [reply]
There is a new method `CDWUSBGadget::OnClassOrVendorRequest()` on the *develop* branch now, which can be overridden to implement class or vendor specific requests on EP0. With this it is possible to implement the `GET MAX LUN` request. Please see the header file *include/circle/usb/gadget/dwusbgadget.h* for details on this method.

A correction: It's the CSD (Card-Specific Data) register/structure, which must be parsed for the SD card capacity.

**Reply by messinm** [reply]
Thanks, will take a look at that.

**Reply by messinm** [reply]
I've written a mass storage gadget class and endpoint class for it, and have started testing. I'm waiting for delivery of a USB serial UART device so I can get log messages while I debug. It works up to the point where my endpoints need to start communicating; seems my endpoints are not working. I'm seeing bulk transfer cancellation "IRP USBD_STATUS: USBD_STATUS_CANCELED (0xc0010000)" after about a twenty second delay. Could be some dumb bugs which I hope to clear up once I get the UART device. But in the meantime, a few questions:

The circle DWUSBEndpoint class seems to be oriented for an endpoint that can work independently of the state of the overall gadget. Seems to fit the case of serial interface and other streaming type of uses. The mass storage device has to send and receive messages of specific lengths (CBW and CSW, etc), and this is sequenced based on the state of the gadget. The state of the gadget changes based on the commands issued by the host. So I wrote my endpoint class to call back to the gadget on transfer complete and the gadget calls into the endpoints to begin a transfer. You may not like this design, please let me know if it seems wrong to you. Since my endpoints seem to not work, maybe I've misunderstood how BeginTransfer and OnTransferComplete work. 

If I call BeginTransfer with a specified length should I expect the OnTransferComplete to be called subsequently with that same length? Is it ok to call BeginTransfer just once and then expect an OnTransferComplete to occur? Is a timeout possible and what happens then? 

Also, could not get the addon/SDCard/sample to work on my Zero 2W. Seems that the CardInit() call is failing. Any ideas of what can be done? I wrote a memory-backed device for testing the gadget, so this isn't holding me up.

Thanks.

**Reply by rsta2** [reply]
I understand, that there are different requirements for the mass-storage gadget, compared to the serial CDC and MIDI gadgets, where the requests can be handled by the endpoint classes. I think, calling back to the gadget class should be OK.

Yes, you have to call `BeginTransfer()` for starting each transfer and you will get a call to `OnTransferComplete()` afterwards, when the transfer completes. For sending data from device to host (TransferDataIn), the length value is the same as given to `BeginTransfer()`. For host to device transfers (TransferDataOut) the length value is less or equal, compared to the value, which was handed over to `BeginTransfer()`, because the host may sent less data. Yes, it's OK to call `BeginTransfer()` and to wait for `OnTransferComplete()` afterwards. A timeout is not implemented.

I tried the *addon/SDCard/sample* again on a RPi Zero 2 W here (HW revision code 902120), and it worked well. You can get more debug output from the SD card driver by setting [SDHOST_DEBUG](https://github.com/rsta2/circle/blob/1884ccf7f630debded2f225f6f931668a6e3264a/addon/SDCard/sdhost.cpp#L46) to 1 before build. Perhaps this helps to find the reason for the problem.

**Reply by messinm** [reply]
I have a working USB mass gadget using a memory-backed device (just for testing).
It's here if you want to take a look: https://github.com/messinm/circle-projects/tree/main/usb-msd-gadget
I'll be trying SD card access next. I tried another SD card and got the addon sample working.

Thanks.

**Reply by rsta2** [reply]
Great. I tested your gadget code on a Raspberry Pi 3 A+, connected to a Linux 6.9.7 host, and the device is detected, how one can see in the dmesg log:

```
[ 1217.597785] usb 1-3: new high-speed USB device number 11 using xhci_hcd
[ 1217.739704] usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x81 has invalid maxpacket 64
[ 1217.739721] usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x2 has invalid maxpacket 64
[ 1217.741434] usb 1-3: New USB device found, idVendor=cafe, idProduct=a4a5, bcdDevice= 1.00
[ 1217.741447] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[ 1217.741452] usb 1-3: Product: Mass Stg Gadget
[ 1217.741457] usb 1-3: Manufacturer: Circle
[ 1217.748034] usb-storage 1-3:1.0: USB Mass Storage device detected
[ 1217.748618] scsi host3: usb-storage 1-3:1.0
[ 1218.769186] scsi 3:0:0:0: Direct-Access     Circle   Mass Storage     000  PQ: 0 ANSI: 4
[ 1218.770039] sd 3:0:0:0: Attached scsi generic sg6 type 0
[ 1218.895151] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
```

Unfortunately the memory device popup does not appear and the device is reset all 30 seconds:

```
[ 1250.375681] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
[ 1250.520166] sd 3:0:0:0: [sdf] 16000 512-byte logical blocks: (8.19 MB/7.81 MiB)
[ 1281.083076] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
[ 1281.231211] sd 3:0:0:0: [sdf] Write Protect is off
[ 1281.231225] sd 3:0:0:0: [sdf] Mode Sense: 00 00 00 00
[ 1311.803538] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
[ 1311.948205] sd 3:0:0:0: [sdf] Asking for cache data failed
[ 1311.948226] sd 3:0:0:0: [sdf] Assuming drive cache: write through
[ 1342.533977] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
[ 1373.244464] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
[ 1373.388830] sdf: detected capacity change from 16000 to 0
[ 1373.388943] sd 3:0:0:0: [sdf] Attached SCSI removable disk
[ 1403.974975] usb 1-3: reset high-speed USB device number 11 using xhci_hcd
```

I tried to open `fdisk /dev/sdf`, but it hangs after the startup message.

I guess, you tested this with Windows and it worked there? Maybe an incompatibility with Linux. Nevertheless this should be a good start.

**Reply by messinm** [reply]
I just now tested with an old Linux version (3.10 Yocto), and it seems to work. Will try to update to a recent kernel and try again.

syslog:

Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: new high-speed USB device number 2 using atmel-ehci
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x81 has invalid maxpacket 64
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x2 has invalid maxpacket 64
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: New USB device found, idVendor=0525, idProduct=a4a5
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=0
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: Product: Mass Stg Gadget
Feb  5 09:24:40 sama5d3_xplained kernel: usb 1-3: Manufacturer: Circle
Feb  5 09:24:40 sama5d3_xplained kernel: usb-storage 1-3:1.0: USB Mass Storage device detected
Feb  5 09:24:40 sama5d3_xplained kernel: usb-storage 1-3:1.0: Quirks match for vid 0525 pid a4a5: 10000
Feb  5 09:24:40 sama5d3_xplained kernel: scsi0 : usb-storage 1-3:1.0
Feb  5 09:24:41 sama5d3_xplained kernel: scsi 0:0:0:0: Direct-Access     Circle   Mass Storage     000  PQ: 0 ANSI: 4
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] 16000 512-byte logical blocks: (8.19 MB/7.81 MiB)
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Write Protect is off
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Mode Sense: 03 00 00 00
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] No Caching mode page present
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Assuming drive cache: write through
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] No Caching mode page present
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Assuming drive cache: write through
Feb  5 09:24:41 sama5d3_xplained kernel:  sda:
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] No Caching mode page present
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Assuming drive cache: write through
Feb  5 09:24:41 sama5d3_xplained kernel: sd 0:0:0:0: [sda] Attached SCSI removable disk

fdisk:

Disk /dev/sda: 8 MB, 8192000 bytes, 16000 sectors
Units = sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disk label type: dos
Disk identifier: 0x6f20736b

This doesn't look like a partition table
Probably you selected the wrong device.

   Device Boot      Start         End      Blocks   Id  System
/dev/sda1   ?   778135908  1919645538   570754815+  72  Unknown
/dev/sda2   ?   168689522  2104717761   968014120   65  Novell Netware 386
/dev/sda3   ?  1869881465  3805909656   968014096   79  Unknown
/dev/sda4   ?           0  3637226495  1818613248    d  Unknown

mount and ls:

root@sama5d3_xplained:~# mkdir /media/usbmsd
root@sama5d3_xplained:~# mount /dev/sda /media/usbmsd
root@sama5d3_xplained:~# ls /media/usbmsd
System Volume Information  circle-usb-msd-gadget-readme.txt
root@sama5d3_xplained:~#

**Reply by messinm** [reply]
I was able to test the msd with kernel version 5.10, and was able to mount it and list the directory. But when doing so, the kernel detected directory corruption. This may be due to the CMemDevice, possibly I inadvertently missed some bits in the disk image.

So I can't recreate the problem you saw, '[ 1373.388830] sdf: detected capacity change from 16000 to 0'. 

I know 5.10 is kind of old now, but it was easy to grab an image with that for the board I have. I'll see if I can get the same version you use.

**Reply by rsta2** [reply]
The [openSUSE Tumbleweed](https://www.opensuse.org/#Tumbleweed) distribution always has a very recent Linux kernel.

**Reply by messinm** [reply]
Ok, I think I have fixed the problem. https://github.com/messinm/circle-projects/tree/main/usb-msd-gadget has been updated. I tested with a 6.6.37 kernel, the device enumerated and I was able to mount and read the demo text file. Next will work on accessing the actual SD card on the Pi.

Thanks.

**Reply by rsta2** [reply]
It works! But I had to change the `wMaxPacketSize` parameter in the endpoint descriptors for the two bulk EPs to 512. Otherwise it does not work and I get the following messages:

```
usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x81 has invalid maxpacket 64
usb 1-3: config 1 interface 0 altsetting 0 bulk endpoint 0x2 has invalid maxpacket 64
```

I tested a little bit, copied something onto the drive, and with this change everything seems to work.

Thanks.

**Reply by messinm** [reply]
I chose maxpacket size of 64 based on this spec: https://www.usb.org/sites/default/files/usbmassbulk_10.pdf
I saw the error messages also but since everything worked I thought it was ok. I'll make the change to 512.

**Reply by rsta2** [reply]
OK. This spec. was written, before USB 2.0 arrived. For USB 1.x with full-speed 64 is OK, but for high-speed 512 should be the right value.

**Reply by messinm** [reply]
I've started on actually using the SD card in the mass storage gadget, and have run into a problem. When the host requests to read sector 0, the CEMMCDevice::Read call hangs. I've traced it to CMMCHost::RequestSync 
(https://github.com/rsta2/circle/blob/749fe99301846217439a40ac7bda547a986db0f7/addon/SDCard/mmchost.cpp#L105)
where it appears that the do-while never exits. Was wondering if there should be an error check condition that also terminates the loop?

**Reply by rsta2** [reply]
Is `CEMMCDevice::Read()` called from the `CDWUSBGadgetEndpoint::OnTransferComplete()` handler? This is not allowed, because `OnTransferComplete()` is called from the IRQ handler and runs at `IRQ_LEVEL`. When the SD card driver tries to trigger an IRQ in this case, it will never receive one, because on `IRQ_LEVEL` IRQs are disabled.

There is only one solution to this: `CEMMCDevice::Read()` must be called at `TASK_LEVEL` from the main loop of the program. Your `CUSBMSDGadget` class needs another method (e.g. `Update()`), which does the calls to the SD cards driver and is manually called from the main loop of the application.

**Reply by messinm** [reply]
Ok, I'll make that change. thanks.

**Reply by messinm** [reply]
https://github.com/messinm/circle-projects has been updated to use the SD card. Tested on Windows and Linux 6.6.37. Requires a 'deviceblocks' arg in cmdline.txt, set to number of SD card blocks.

Do you plan to work on SD card capacity (from CSD)? if not I may give it a try.

Thanks.

**Reply by rsta2** [reply]
It works well here with Linux 6.9.9. Good job! Yes, I will add the `CEMMCDevice::GetSize()` function in the next days with the info from the CSD register.

**Reply by rsta2** [reply]
`CEMMCDevice::GetSize()` is implemented on the *develop* branch now. It does work with SD cards only, not with the `USE_EMBEDDED_MMC_CM` option.

**Reply by messinm** [reply]
Thanks for that quick update. I made some minor changes to my code and everything looks good.

Question about synchronization - There's no reason for anyone to call the Update() function from multiple cores, but should I add a spinlock as a precaution? Also, to make sure I understand how that would work (even though may not be needed), the spinlock would need to be constructed with TASK_LEVEL target? Since it calls IO routines, must stay at TASK_LEVEL when the lock is acquired, right?

**Reply by rsta2** [reply]
You are welcome. Thanks for info! Your understanding of the spinlock is right. If it would be constructed with IRQ_LEVEL target, the execution level would rise to IRQ_LEVEL on acquiring the spinlock, and the SD card access wouldn't work as before. But I think, a spinlock is not necessary here, because the developer has control, where `Update()` is called, and acquiring and releasing a spinlock creates some (small) overhead.

**Reply by messinm** [reply]
Do you want me to prepare a pull request for the mass storage gadget, or will you integrate this yourself?

Thanks.

**Reply by rsta2** [reply]
Thank you for contributing your MSD gadget driver! I will integrate it from your known repository into Circle.

**Reply by messinm** [reply]
You're welcome, and thanks for circle!

**Reply by rsta2** [reply]
The integrated USB MSD gadget driver is on the branch *usb-msd-gadget* with the test program in *test/usb-msd-gadget*. I did some small modifications and removed the memory device support in the test, because it is not necessary any more, with the SD card supported. You have to add the option `logdev=ttyS1` to the file *cmdline.txt* on the SD card, to get the log output via the serial interface. I tested it on a RPi 3 A+ with a Linux 6.9.9 host so far.

Can you please have a look, if this is OK so far. If you have modifications, please send a PR for the branch *usb-msd-gadget*. Probably I will also have another look again. After completion this branch will be merged to the *develop* branch to be part of the next release.

**Reply by messinm** [reply]
It looks good to me, thanks.

**Reply by rsta2** [reply]
The USB MSD gadget driver and test has been merged to the *develop* branch now. The branch *usb-msd-gadget* has been removed. Thanks again for your work!

**Comment by rsta2** [comment]
I have made the `bDeviceClass` modification on the *develop* branch and added support for the system option `NO_BUSY_WAIT` to *dwusbgadget.cpp*. When this option is defined, the (longer) delays are replaced with sleep calls. You have to define this option, because this needs the scheduler in the system, which is not always available. `NO_BUSY_WAIT` was already used in other drivers (e.g. USB, EMMC).

@messinm Thanks for reporting this, and @sebastienNEC Thanks for testing the recommended change and for clarifying the situation.

**Comment by messinm** [comment]
You're welcome!
